### PR TITLE
Add types for unlock-folders component

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

This adds the types for unlock-folders, so we should be able to work on hooking up actions (making it functional) and creating the components (making it visible) in parallel.

### A rough idea on a suggested way to create components:
Follow @gabriel's example with the component tracker and make a version of the component with each mock state. Then we have one place to see all the versions of the component, and it'll be helpful in styling each section without having to reproduce the state changes to get there.

### CODE REVIEW

File: shared/constants/types/flux.js [hunk 1]
```diff
@@ -16,4 +16,4 @@ export type AsyncAction = (dispatch: Dispatch, getState: GetState) => ?Promise
 export type Dispatch = (action: TypedAction | AsyncAction) => ?Promise
 
 export type TypedAsyncAction<A> = (dispatch: TypedDispatch<A>, getState: GetState) => ?Promise
-export type TypedDispatch<A> = (action: A | TypedAsyncAction<A>) => ?Promise
+export type TypedDispatch<A> = (action: TypedAsyncAction<A> | A) => ?Promise
```
Comment (MarcoPolo) [inline shared/constants/types/flux.js]: I had to do this switch or flow would give me errors when I passed in a `TypedAsyncAction` into a `TypedDispatch` only sometimes. Really weird.

File: shared/constants/unlock-folders.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/* @flow */
+
+import HiddenString from '../util/hidden-string'
+import type {TypedAction} from '../constants/types/flux'
+import type {DeviceID} from '../constants/types/flow-types'
+
+export type Device = {
+  type: 'desktop' | 'mobile',
```
Comment (MarcoPolo) [inline shared/constants/unlock-folders.js]: Not sure if there are other types here. `common.avdl` isn't helpful it just says this is a string.

File: shared/reducers/unlock-folders.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+import * as Constants from '../constants/unlock-folders'
+import HiddenString from '../util/hidden-string'
+
+import type {UnlockFolderActions, Device} from '../constants/unlock-folders'
+
+export type State = {
+  phase: 'dead' | 'promptOtherDevice' | 'paperKeyInput' | 'success',
+  devices: ?Array<Device>,
+  paperkey: ?HiddenString,
```
Comment (MarcoPolo) [inline shared/reducers/unlock-folders.js]: Ideally we wouldn't use this, so whoever does the functional thing should avoid it

File: shared/actions/unlock-folders.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import type {RouteAppend} from '../constants/router'
+import type {TypedAsyncAction} from '../constants/types/flux'
+import type {LoadDevices, ToPaperKeyInput, CheckPaperKey, Finish} from '../constants/unlock-folders'
+
+// TODO this won't work in a remote component until we have a synchronous way to get the latest state in a remote component.
+
+function nextPhase (): TypedAsyncAction<RouteAppend> {
+  return (dispatch, getState) => {
+    // TODO careful here since this will not be sync on a remote component!
+    const phase: string = getState().unlockFolders.phase
```
Comment (chrisnojima) [inline shared/actions/unlock-folders.js]: i find this model problematic. we're doing this in signup and i think its strange the the value in the store is disconnected from what we see. maybe i'm doing something wrong but what i see is phase = '1' and its not on screen '1' until some action decides to call this to make the ui reflect the store. am i misunderstanding this

File: shared/constants/unlock-folders.js (hunk 2 above)
Comment (chrisnojima) [inline shared/constants/unlock-folders.js]: those are the only 2 types, its in constants.go which isn't generated. we should just manually add this to our generated types in flow-types actually

File: shared/constants/unlock-folders.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export const loadDevices = 'unlockFolders:loadDevices'
+export type LoadDevices = TypedAction<'unlockFolders:loadDevices', {devices: Array<Device>}, {}>
+
+// transistions to the next paper key phase
+export const toPaperKeyInput = 'unlockFolders:toPaperKeyInput'
+export type ToPaperKeyInput = TypedAction<'unlockFolders:toPaperKeyInput', {}, {}>
+
+// TODO: not sure if we even need this, we might be able to do it all without having to store the paperkey
```
Comment (chrisnojima) [inline shared/constants/unlock-folders.js]: we can avoid having the keys travel through redux entirely using the dynamic routes with a onSumbit being plumbed through directly to a response.result() call

File: shared/reducers/unlock-folders.js (hunk 3 above)
Comment (chrisnojima) [inline shared/reducers/unlock-folders.js]: i think its fine for it to be in the state, it shouldn't be in the props

File: shared/actions/unlock-folders.js (hunk 4 above)
Comment (MarcoPolo) [inline shared/actions/unlock-folders.js]: So this `nextPhase` is just to have the router reflect the phase of the flow.

It's only purpose is to emit an action to the router (like `routeAppend`) so the route matches the phase.

So it isn't that the UI doesn't match what the phase is until this action is fired, more like the UI _could_ match the router state, and the router state doesn't show the next phase until this action is fired.

**but**, I've thought more about this, and I don't think we should be doing routing (using our Router thing) in remote components. It just adds a bunch of complexity for very little gain.

That means we won't need this `nextPhase` action and the component will reflect it's phase by looking at the phase value in the store.

File: shared/constants/unlock-folders.js (hunk 5 above)
Comment (MarcoPolo) [inline shared/constants/unlock-folders.js]: agree

File: shared/actions/unlock-folders.js (hunk 4 above)
Comment (chrisnojima) [inline shared/actions/unlock-folders.js]: sounds good to me

**Comment by chrisnojima** [comment]
:+1:

**Pull request merged**